Add Middleware for parsing body data based on content type.

markstory opened the pull request:
This adds a middleware based option for the automatic request body parsing that RequestComponentHelper provides. A middleware option is required to enable other middleware (like authentication) access to JSON/XML based request payloads.

Refs cakephp/authentication#185

markstory committed:
Add Middleware for parsing body data based on content type.

This adds a middleware based option for the automatic request body
parsing that RequestComponentHelper provides. A middleware option is
required to enable other middleware (like authentication) access to
JSON/XML based request payloads.

markstory committed:
Add tests for unknown type scenario.

markstory committed:
Add test for charset removal.

lorenzo commented:
What happens if both this middleware and the component are present in a json payload request?

markstory commented:
> What happens if both this middleware and the component are present in a json payload request?

The request body will be parsed twice. The request body payload is inside a buffered stream which will be read by both layers.

Longer term, I think RequestHandlerComponent should go away, and be replaced by a component that is focused on layout/view switching as all the other responsibilities of this component have been deprecated - `beforeRedirect` handling, or moved to other places - `request->is()`, and this change.

lorenzo commented:
@markstory Can we set a custom arg to the request the RequestHandlerComponent can read and avoid doing the same work twice? I don't think a lot of people will remove the component after migrating to 3.6

markstory commented:
@lorenzo We could skip parsing input if `getParsedBody()` returns a non-empty value. That would let us not have to add an additional config option.

lorenzo commented:
@markstory That sounds good to me!

chrisShick commented:
@markstory If we were to remove the RequestHandler, would we also not need another middleware for response as ?

markstory committed:
Don't double parse request body.

If the request only contains a parsed body don't re-parse it.
This lets the middleware and component play nicer together.

markstory commented:
@lorenzo Updated.
